File: solutions/go/anagram/anagram.go
package anagram

import (
	"sort"
	"strings"
)

func sorted(word string) string {
	chars := []rune(word)
	sort.Slice(chars, func(i, j int) bool {
		return chars[i] < chars[j]
	})
	return string(chars)
}

// Detect returns all candidates that are anagrams of, but not equal to, 'subject'.
func Detect(subject string, candidates []string) (anagrams []string) {
	base := strings.ToLower(subject)
	base_letters := sorted(base)

	for _, candidate := range candidates {
		if len(base) != len(candidate) {
			continue
		}
		word := strings.ToLower(candidate)
		if base == word {
			continue
		}
		word_letters := sorted(word)
		if base_letters == word_letters {
			anagrams = append(anagrams, candidate)
		}
	}
	return
}

File: solutions/r/anagram/anagram.R
# Returns all candidates that are anagrams of, but not equal to, 'subject'.
anagram <- function(subject, candidates) {
    sorted <- function(input) {
        sort(strsplit(input, "")[[1]])
    }

    base <- toupper(subject)
    is_anagram <- sapply(candidates, function(word) {
        word <- toupper(word)
        nchar(word) == nchar(base) &&
            word != base &&
            all(sorted(word) == sorted(base))
    })

    if (any(is_anagram)) {
        candidates[is_anagram]
    } else {
        c()
    }
}
